sdk/core/core/src/az_json_data.c

```diff
@@ -0,0 +1,220 @@
+    AZ_RETURN_IF_FAILED(
+        az_mut_span_move(az_mut_span_drop(builder->buffer, new_size), data.span, &result));
+  }
+  builder->buffer.size = new_size;
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result _az_span_builder_top_array_revert(
+    az_span_builder * const builder,
+    _az_type const item_type,
+    size_t const new_size,
+    void ** const out_array_begin,
+    size_t * const out_array_size) {
+  AZ_CONTRACT_ARG_NOT_NULL(builder);
+  AZ_CONTRACT_ARG_NOT_NULL(out_array_begin);
+  AZ_CONTRACT_ARG_NOT_NULL(out_array_size);
+
+  // 3. restore buffer size
+  size_t const offset = builder->buffer.size;
+  builder->buffer.size = new_size;
+  az_mut_span const buffer = builder->buffer;
+
+  // 1. revert the array.
+  size_t const item_count = (new_size - offset) / item_type.size;
+  for (size_t i = 0; i < item_count / 2; ++i) {
+    az_mut_span const a
+        = az_mut_span_take(az_mut_span_drop(buffer, offset + i * item_type.size), item_type.size);
+    az_mut_span const b = az_mut_span_take(
+        az_mut_span_drop(buffer, offset + (item_count - i - 1) * item_type.size), item_type.size);
+    az_mut_span_swap(a, b);
+  }
+
+  // 2. move it to front.
+  az_data const data = {
+    .span
+    = az_span_take(az_span_drop(az_mut_span_to_span(buffer), offset), item_count * item_type.size),
+    .align = item_type.align,
+  };
+  AZ_RETURN_IF_FAILED(_az_span_builder_aligned_append(builder, data, out_array_begin));
+  *out_array_size = item_count;
+
+  return AZ_OK;
+}
+
+AZ_INLINE AZ_NODISCARD az_result _az_span_builder_append_json_data(
+    az_span_builder * const builder,
+    az_json_data const data,
+    az_json_data const ** const out) {
+  uint8_t * p = { 0 };
+  AZ_RETURN_IF_FAILED(_az_span_builder_aligned_append(builder, AZ_DATA(data), &p));
+  *out = (az_json_data const *)p;
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result _az_span_builder_write_json_string(
+    az_span_builder * const builder,
+    az_span const json_string,
+    az_span * const out) {
+  // TODO: escape symbols.
+  *out = (az_span){ .begin = builder->buffer.begin + builder->length, json_string.size };
+  return az_span_builder_append(builder, json_string);
+}
+
+AZ_NODISCARD az_result _az_span_builder_write_json_value(
+    az_span_builder * const builder,
+    az_json_parser * const parser,
+    az_json_token const token,
+    az_json_data * const out) {
+  AZ_CONTRACT_ARG_NOT_NULL(builder);
+  AZ_CONTRACT_ARG_NOT_NULL(parser);
+  AZ_CONTRACT_ARG_NOT_NULL(out);
+
+  switch (token.kind) {
+    case AZ_JSON_TOKEN_NULL: {
+      *out = az_json_data_null();
+      break;
+    }
+    case AZ_JSON_TOKEN_BOOLEAN: {
+      *out = az_json_data_boolean(token.data.boolean);
+      break;
+    }
+    case AZ_JSON_TOKEN_NUMBER: {
+      *out = az_json_data_number(token.data.number);
+      break;
+    }
+    case AZ_JSON_TOKEN_STRING: {
+      az_span s = { 0 };
+      AZ_RETURN_IF_FAILED(_az_span_builder_write_json_string(builder, token.data.string, &s));
```

Comment:
again I want to see tests that the builder structure is valid after these failures.